### PR TITLE
Add smoothed L1 loss and enhancements to cross entropy loss 

### PR DESCRIPTION
**Pull request by Wei-Sheng-Wang**
## Proposed changes

1. Added smoothed l1 loss
2. Added weights and label smoothing to cross entropy loss

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Review by awni — CHANGES_REQUESTED** [review]
Awesome! This is great! I left a few comments.

Also I noticed you changed the reductions from 'none' to `mean'. What was the motivation there? We have been intentionally keeping them at 'none' for compatibility reasons, but I'm wondering your motivation for changing it?

File: python/mlx/nn/losses.py [hunk 1]
```diff
@@ -195,6 +220,32 @@ def kl_div_loss(
...
+          ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'mean'``.
+
+    Returns:
+        mx.array: The calculated loss. It will be a single element array if reduction is 'mean' or 'sum',
+                     or the same shape as predictions and targets if reduction is 'none'.
+    """
+
+    diff = mx.abs(predictions - targets)
+    loss = mx.where(
+        diff < beta, 0.5 * mx.square(predictions - targets) / beta, diff - 0.5 * beta
```
[the remark this answers is not in the document]
Comment (Wei-Sheng-Wang) [reply python/mlx/nn/losses.py:243]: do you mean using 
```mx.maximum(predictions - targets, targets - predictions) ```
instead of
```mx.abs(predictions - targets)``` ? 

Could you explain why mx.maximum would be faster?

[the remark this answers is not in the document]
Comment (awni) [reply python/mlx/nn/losses.py:243]: No that's not what I mean sorry for not being clear:
```
 diff = predictions - targets
 loss = mx.maximum(0.5 * mx.square(diff) / beta, mx.abs(diff) - 0.5 * beta)
```

[the remark this answers is not in the document]
Comment (Wei-Sheng-Wang) [reply python/mlx/nn/losses.py:243]: The smooth L1 loss is defined according to the piecewise function below, I don't think taking the maximum would be correct here. What do you think ? 

$$
\text{loss} = 
\begin{cases} 
\frac{0.5 \times (predictions - targets)^2}{beta}, & \text{if } diff < beta \\
diff - 0.5 \times beta, & \text{otherwise}
\end{cases}
$$

**Comment by Wei-Sheng-Wang** [comment]
For cross entropy, I thought it makes sense to set the reduction to mean since it's most commonly used. Also I've noticed a couple of things 

```
def l1_loss(
    predictions: mx.array, targets: mx.array, reduction: str = "none"
) -> mx.array:

    loss = mx.mean(mx.abs(predictions - targets))
    return _reduce(loss, reduction)


def mse_loss(
    predictions: mx.array, targets: mx.array, reduction: str = "none"
) -> mx.array:

    loss = mx.square(predictions - targets)
    return _reduce(loss, reduction)

```

Q1. For L1 loss, why is the mean taken directly in the loss calculation instead of letting _reduce handle it, i.e.
```
 loss = mx.abs(predictions - targets)
 return _reduce(loss, reduction='mean')
```
Q2. For MSE loss, why didn't you take the mean similar to L1 loss? i.e.
```
loss = mx.mean(mx.square(predictions - targets))
```

**Comment by awni** [comment]
> Q1. For L1 loss, why is the mean taken directly in the loss calculation instead of letting _reduce handle it, i.e.

Actually that's a bug for L1, good catch! I think for L1 and MSE we could switch the defaults to `mean` since that is much more common. For the others, please keep them as `none` for now. We may switch later but that needs a bit more thought.

File: python/mlx/nn/losses.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (awni) [reply python/mlx/nn/losses.py:243]: Sorry that's my bad.. disregard the maximum suggestion. Storing the difference is still a good idea.

```
    diff = predictions - targets
   abs_diff = mx.abs(diff)
    loss = mx.where(
        abs_diff < beta, 0.5 * mx.square(diff) / beta, abs_diff - 0.5 * beta
```

**Comment by Wei-Sheng-Wang** [comment]
Thanks for the feedback! I've made the necessary changes:

1. Update the docs to use double backticks for default values
2. Set reduction defaults to 'mean' for l1 and l2 loss
3. Delegate the reduction in l1 loss to _reduce
4. Update the docs for smooth l1 loss

Could you please check if I missed out on anything?

**Comment by Wei-Sheng-Wang** [comment]
Thanks of the feedback ! Based on your suggestions, I've made the following updates:

1. Use double backticks to indicate default values in docs
2. Included a detailed explanation for smooth l1 loss function
3. Used a temporary variable to avoid re-computation, i.e. `diff = predictions - targets`
4. Delegated reduction functionality in l1 loss function to `_reduce` function
5. Change the default value of `reduction` argument to 'mean' in both the l1 and l2 loss functions

Could you please review these changes and let me know if there's anything I might have missed?

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -15,24 +15,57 @@ def decorator(klass):
...
+    if label_smoothing > 0:
+        # Adjust the true class score with label smoothing
+        adjusted_score = (1 - label_smoothing) * score
+
+        # Calculate the mean logit across the classes for smoothed loss
+        mean_logits = logits.mean(axis=axis)
+        smoothed_loss = -mean_logits * label_smoothing
+
+        # Combine the adjusted score and smoothed loss with the logsumexp logits
+        loss = logsumexp_logits - adjusted_score + smoothed_loss
```
Comment (awni) [inline python/mlx/nn/losses.py:58]: Are you basing this off of a reference implementation? Just curious as I see label smoothing defined as `(1-l)` for the ground truth and `l / (N -1)` for the remaining labels. where `l` is the label smoothing parameter. 

But here the label smoothing parameter gives you `1 - l + l/N` for the ground truth so it's a slightly different parameterization. I like the simplicity of implementation, but I'm just wondering what's standard and if we should follow that?

**Comment by awni** [comment]
@Wei-Sheng-Wang I had a question about the label smoothing implementation. Otherwise this is looking great to me!

File: python/mlx/nn/losses.py (hunk 2 above)
Comment (Wei-Sheng-Wang) [reply python/mlx/nn/losses.py:58]: There are many implementations out there I believe. I referred to some of them [here](https://stackoverflow.com/questions/55681502/label-smoothing-in-pytorch), though it's not an exact replica of any particular implementation. I kinda adapted them to optimize computation by reusing `logsumexp` and `score`.

From my understanding, label smoothing reduces the confidence of the correct class $y_{c} = 1$ by $l$ and redistributes $l$ evenly among all classes, which includes the correct class too. Hence, the smoothed label for the correct class becomes $1 - l + \frac{l}{N}$, and each incorrect class gets $\frac{l}{N}$, where $N$ is the total number of classes.

The standard cross-entropy without label smoothing is given as:

$$
\begin{aligned}
CE(\textbf{x},\textbf{y}) &= - \sum_{k=1}^{N} y_k \log(p(x_k)) \\
&= - \sum_{k=1}^{N} y_k \log\left(\frac{\exp(x_k)}{\sum_{i} \exp(x_i)}\right) \\
&= - \sum_{k=1}^{N} y_k \left( x_k - \log\left( \sum_{i=1}^{N} \exp(x_i) \right) \right) \\
&=  \log\left( \sum_{i=1}^{N} \exp(x_i) \right) - \sum_{k=1}^{N} y_k x_k 
\end{aligned}
$$

Including label smoothing, the term $\sum_{k} y_k x_k$ is recalculated as:

$$
\begin{aligned}
\sum_{k=1}^{N} y_k x_k
&= \left( (1 - l) + \frac{l}{N} \right) x_c + \sum_{k \neq c} \frac{l}{N} x_k \\
&= (1 - l) x_c + \frac{l}{N} \sum_{k=1}^{N} x_k
\end{aligned}
$$

The contributions from the modified labels sum to 1:

- For the correct class $c$, the contribution is $(1 - l) + \frac{l}{N}$.
- For each of the $N - 1$ incorrect classes, the contribution is $\frac{l}{N}$. Summing this over all $N - 1$ classes gives $(N - 1) \times \frac{l}{N}$.

The total sum is then:

$$
(1 - l) + \frac{l}{N} + (N - 1) \times \frac{l}{N} = 1
$$

This shows that after label smoothing, the sum of the weighted labels remains 1, maintaining the property of a probability distribution.

I hope that answers your question!

Comment (awni) [reply python/mlx/nn/losses.py:58]: What you are saying makes perfect sense, and thanks for the details. I had no doubt it was a valid distribution you are computing, my only point is that an alternative parameterization is to have the ground truth get a weight of $(1-l)$ instead of $(1-l) + /frac{1}{N}$. In that case the other labels get a weight of $\frac{l}{N-1}$.

That said, what you have is definitely cleaner/simpler to implement so I'm in favor of keeping it the way it is.

Comment (Wei-Sheng-Wang) [reply python/mlx/nn/losses.py:58]: Ah, I see where you're coming from. Sorry, I misunderstood. You're right that there's an alternative that uses $1-l$ for the ground truth and $\frac{l}{N-1}$ for the other incorrect labels. I'm not really sure which is standard as there are many implementations that use either one. In this regard, I'm following [Nvidia](https://github.com/NVIDIA/DeepLearningExamples/blob/8d8b21a933fff3defb692e0527fca15532da5dc6/PyTorch/Classification/ConvNets/image_classification/smoothing.py#L18) in using $1-l+\frac{l}{N}$ for ground truth.

**Review by awni — APPROVED** [review]
Thanks for adding this!! 🚀

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -4,163 +4,156 @@
...
+        >>> inputs = mx.array([0.105361, 0.223144, 1.20397, 0.916291])
         >>> targets = mx.array([0, 0, 1, 1])
-        >>> loss = nn.losses.binary_cross_entropy(inputs, targets)
+        >>> loss = nn.losses.binary_cross_entropy(inputs, targets, "mean")
         >>> loss
-        array([0.612192])
+        array([0.612192], dtype=float32)
     """
-    loss = -targets * mx.log(inputs) - (1 - targets) * mx.log(1 - inputs)
+    loss = mx.logaddexp(0.0, logits) - targets * logits
```
Comment (sciencecw) [inline python/mlx/nn/losses.py:84]: Why use `mx.logaddexp(0.0, logits)` for softmax and not just write it out to `log(1+exp(logits))`?
